### PR TITLE
[web] [fix] Cache resource data only if the fetching succeed

### LINKED ISSUE #99671: Flutter web service worker caches error response
`flutter_service_worker.js` seems to cache the response even if the response is 4XX/5XX error.
https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/build_system/targets/web.dart#L669

And, the service worker caches the response during the resource's hash is identical.
https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/build_system/targets/web.dart#L616

As a result, the service worker still uses the cache of the error response after the error is recovered, because the target resource's hash is not changed.
Is it expected behavior? Or, should the cacheing strategy be improved?

## Steps to Reproduce

1. Request to some static resource e.g. `.js` file
1. The request failed with an error
     - In my case it is 401 Basic Authentication failure, but I think the same thing will occur in other situation e.g. CDN is down.
1. The error response is cached by flutter service worker
1. Recovered from the error
1. Request to the target resource again

**Expected results:** 

The resource file is loaded successfully

**Actual results:** 

The cached error is still returned by the service worker

### PR DESCRIPTION
**Pull request by dacianf**
Fix for issue #99671 

Caching the resulting resource is done regardless of whether the call was successful. This results in caching both successful and failed results. 
This PR changes the logic to only cache successful results. 

Closes #99671

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by dacianf** [comment]
@ditman When you have time, I would appreciate your review.

File: dev/bots/service_worker_test.dart [hunk 1]
```diff
@@ -78,14 +82,40 @@ Future<void> _rebuildApp({ required int version, required ServiceWorkerTestType
...
+}
+
+Future<void> _waitForAppToLoad(
+    Map<String, int> waitForCounts,
+    Map<String, int> requestedPathCounts,
+    AppServer? server
+    ) async {
+  print('Waiting for app to load $waitForCounts');
+  await Future.any(<Future<Object?>>[
+        () async {
```
Comment (yjbanov) [inline dev/bots/service_worker_test.dart:108]: Fix indent (-4 spaces)

File: dev/bots/service_worker_test.dart [hunk 2]
```diff
@@ -78,14 +82,40 @@ Future<void> _rebuildApp({ required int version, required ServiceWorkerTestType
...
     environment: <String, String>{
       'FLUTTER_WEB': 'true',
     },
   );
 }
 
+void _expectRequestCounts(
+    Map<String, int> expectedCounts,
+    Map<String, int> requestedPathCounts,
+    ) {
```
[the remark this answers is not in the document]
Comment (dacianf) [reply dev/bots/service_worker_test.dart]: changed

File: dev/bots/service_worker_test.dart [hunk 3]
```diff
@@ -78,14 +82,40 @@ Future<void> _rebuildApp({ required int version, required ServiceWorkerTestType
...
+  expect(requestedPathCounts, expectedCounts);
+  requestedPathCounts.clear();
+}
+
+Future<void> _waitForAppToLoad(
+    Map<String, int> waitForCounts,
+    Map<String, int> requestedPathCounts,
+    AppServer? server
+    ) async {
```
[the remark this answers is not in the document]
Comment (dacianf) [reply dev/bots/service_worker_test.dart]: changed

File: dev/bots/service_worker_test.dart (hunk 1 above)
Comment (dacianf) [reply dev/bots/service_worker_test.dart]: changed

**Review by yjbanov — APPROVED** [review]
lgtm

**Review by kevmoo — APPROVED** [review]

**Pull request merged**